Issue 23: Differences between Internal and standard should be listed

pjohnsoninfusion opened the pull request:
- add mocking for git and jira for dev/testing.  enabled using config flag 'devMode=true'.  output to an 'objects' directory as a sibling directory to report.directory
    - uses xstream, added to pom.  outputs in XML format
    - FileUtils - utility classes for saving/reading objects
    - XstreamSerializer - utilty class to serialize/deserialize objects, ** not type checked
    - JiraConnectorMock - manages serialized objects, reads new data in if no objects exist
- add support for separate internal and external report templates (see Readme.md)
- add separation in internal report between external and internal issues
    - ModelViewList - enum to store view levels (internal/external)
- updated default report.ftl to use internal and external data views
- refactoring
    - moved builders into separate classes

pjohnsoninfusion committed:
- add mocking for git and jira for dev/testing.  enabled using config flag 'devMode=true'.  output to an 'objects' directory as a sibling directory to report.directory
    - uses xstream, added to pom.  outputs in XML format
    - FileUtils - utility classes for saving/reading objects
    - XstreamSerializer - utilty class to serialize/deserialize objects, ** not type checked
    - JiraConnectorMock - manages serialized objects, reads new data in if no objects exist
- add support for separate internal and external report templates (see Readme.md)
- add separation in internal report between external and internal issues
    - ModelViewList - enum to store view levels (internal/external)
- updated default report.ftl to use internal and external data views
- refactoring
    - moved builders into separate classes

karpikpl commented on src/main/java/com/infusion/relnotesgen/Configuration.java:
```diff
@@ -50,12 +50,15 @@
     static final String ISSUE_SORT_PRIORITY = "issue.sort.priority";
     static final String REPORT_DIRECTORY = "report.directory";
     static final String REPORT_TEMPLATE = "report.template";
```
what's the difference between `REPORT_TEMPLATE` and `REPORT_EXTERNAL_TEMPLATE`?
Is the first one still needed?

karpikpl commented on README.md:
````diff
@@ -67,7 +67,8 @@ Check ```com.infusion.relnotesgen.MainITTest``` for appropriate use case.
 | report.clientFacingFilters			| -clientFacingFilters 				| Indicates fields which indicate issue should be included in client facing report | Requirement VA ID,Defect_Id |
-| report.template			| -reportTemplate 				| Path to external template for release notes. Freemarker is used as template engine. Variables provided in context: \$\{issues\} - map of issues where key is issue type and value list of issues; \$\{jiraUrl\} - url to jira; \$\{version\} - version for which release notes are generated | C:/releaseNotes/template.ftl |
+| report.internalTemplate			| -reportInternalTemplate 				| Path to template for internal only release notes. Freemarker is used as template engine. Variables provided in context: \$\{issues\} - map of issues where key is issue type and value list of issues; \$\{jiraUrl\} - url to jira; \$\{version\} - version for which release notes are generated | C:/releaseNotes/internalTemplate.ftl |
+| report.externalTemplate			| -reportExternalTemplate 				| Path to template for external release notes. Freemarker is used as template engine. Variables provided in context: \$\{issues\} - map of issues where key is issue type and value list of issues; \$\{jiraUrl\} - url to jira; \$\{version\} - version for which release notes are generated | C:/releaseNotes/externalTemplate.ftl |
````
can this be relative URL or has to be absolute?

karpikpl reviewed (changes requested):
I got a NPE when template path was missing in 

00:09:55.132 INFO  c.i.r.l.ReleaseNotesLogger - HA-10443 is valid
00:09:55.132 DEBUG c.i.r.l.ReleaseNotesLogger - HA-10446 is valid, at least one FixVersions for the Jira Issue is included in conguration jira.fixVersions
00:09:55.132 INFO  c.i.r.l.ReleaseNotesLogger - HA-10446 is valid
00:09:55.132 INFO  c.i.r.l.ReleaseNotesLogger - Validating issues
00:09:55.147 INFO  c.i.r.l.ReleaseNotesLogger - Getting known issues for query: project = HA AND status not in ("PO Review", completed, removed, "Configuration Done") AND type not in (subTaskIssueTypes(), test) AND (fixVersion is EMPTY OR fixVersion in unreleasedVersions()) AND type = bug AND ("Epic Link" is EMPTY OR "Epic Link" not in (HA-8832, HA-8843, HA-6919, HA-8834, HA-8833)) AND ("Discovered In" is EMPTY OR "Discovered In" not in (R3.3, R4)) ORDER BY cf[11801] DESC, status DESC
00:10:01.837 INFO  c.i.r.l.ReleaseNotesLogger - Fetching issues from JIRA completed. 116 issues fetched.
00:10:01.919 INFO  c.i.r.l.ReleaseNotesLogger - Using template report.ftl
Exception in thread "main" java.lang.NullPointerException
	at freemarker.template.Configuration.setDirectoryForTemplateLoading(Configuration.java:436)
	at com.infusion.relnotesgen.ReleaseNotesReportGenerator.generateTemplateNameAndInitialize(ReleaseNotesReportGenerator.java:74)
	at com.infusion.relnotesgen.ReleaseNotesReportGenerator.getFreemarkerTemplate(ReleaseNotesReportGenerator.java:46)
	at com.infusion.relnotesgen.ReleaseNotesReportGenerator.<init>(ReleaseNotesReportGenerator.java:26)
	at com.infusion.relnotesgen.Main.generateReleaseNotesFile(Main.java:99)
	at com.infusion.relnotesgen.Main.generateReleaseNotes(Main.java:79)
	at com.infusion.relnotesgen.Main.main(Main.java:35)

pjohnsoninfusion committed:
resolving merge conflicts

pjohnsoninfusion committed:
- fixed NPE when template path missing
- removed 'REPORT_TEMPLATE' from Configuration
- added support for relative template filenames (can be explicit or relative to src/main/resources)

pjohnsoninfusion commented:
Commit 8984318 should resolve the issues reported.

- fixed NPE when template path missing
- removed 'REPORT_TEMPLATE' from Configuration
- added support for relative template filenames (can be explicit or relative to src/main/resources)

karpikpl commented:
I'm getting NPE when running the generator.
params:
`java -jar target/release-notes-generator-1.0-SNAPSHOT.jar -configurationFilePath ../configuration.properties -tag2 1.3.0_78-RELEASE -tag1 1.3.0_76-RELEASE`

config:
```
jira.fixVersions=R3.2 - Harmony 1.3.0_78
report.directory=./
report.externalTemplate=./src/main/resources/report.ftl
report.internalTemplate=./src/main/resources/report.ftl
report.clientFacingFilters=Requirement VA ID,Defect_Id
```

pjohnsoninfusion committed:
adding support for
1. absolute windows/linux (C:/releaseNotes/internalTemplate.ftl)
2. relative to pwd (./src/main/resources/report.ftl)
3. path within src/main/resources (report.ftl)

pjohnsoninfusion commented:
That should do it. Added support for:
1. absolute windows/linux (C:/releaseNotes/internalTemplate.ftl or /home/releaseNotes/internalTemplate.ftl)
2. relative to pwd (./src/main/resources/report.ftl)
3. path within src/main/resources (report.ftl)

pjohnsoninfusion committed:
rewording to make it clearer
